src/bin/genemichaels.rs

```diff
@@ -250,14 +330,147 @@ fn main() {
                 Ok(_) => { },
                 Err(e) => {
                     if !args.quiet {
-                        eprintln!("Error formatting {}: {:?}", &file.to_string_lossy(), e);
+                        print_error_text();
+                        eprintln!("formatting {}: {:?}", &file.to_string_lossy(), e);
                     }
                     failed = true;
                 },
             };
         }
+        eprintln!(
+            "\x1B[1;32m    Finished\x1B[0;22m workspace formatted successfully in {:.2}s",
+            time::Instant::now().duration_since(inst).as_secs_f64()
+        );
         if failed {
             process::exit(1);
         }
     }
 }
+
+fn process_cargo_toml(
+    path: PathBuf,
+    manifest: Manifest,
+    thread_count: Option<usize>,
+    config: FormatConfig,
+) -> Result<()> {
+    let mut dirs = Vec::from([]);
+    for bin in manifest.bin.into_iter().flatten() {
+        if let Some(bin_path) = bin.path {
+            let bin_path =
```

Comment:
Okay cool, so doing `path.join(dir)` at the end resolved that one issue.  I think it would be better to do it here
1. You can avoid the `PathBuf` parse error handling and thus some extra chaining (you don't need `to_path_buf` either I think)
2. There's also `path.join` for each of the default dirs (`"src", "benches", etc below).  Right now they're no-ops probably because it's a join of absolute paths, but it would be better to avoid duplication.  Doing `path.join` on each added path as they're added rather than at the end will fix that.

FWIW I don't think nicely handling the parent-is-none issue is important here, I'd just unwrap.  It shouldn't be possible for that to normally occur, so if it does it's probably a bug and a panic is reasonable.

If you do all that I think this will be simplified to
```
if let Some(bin_path) = bin.path {
    dirs.push(path.join(bin_path).parent().unwrap());
}
```

Reply:
Looks good